examples/ex0.cpp

```diff
@@ -0,0 +1,80 @@
+//                                MFEM Example 0
+//
+// Compile with: make ex0
+//
+// Sample runs:  ex0
+//               ex0 -m ../data/fichera.mesh
+//               ex0 -m ../data/square-disc.mesh -o 2
+//
+// Description: This example code demonstrates the most basic usage of MFEM to
+//              define a simple finite element discretization of the Laplace
+//              problem -Delta u = 1 with zero Dirichlet boundary conditions.
+//              General 2D/3D mesh files and finite element polynomial degrees
+//              can be specified by command line options.
+
+#include "mfem.hpp"
+#include <fstream>
+#include <iostream>
+
+using namespace std;
+using namespace mfem;
+
+int main(int argc, char *argv[])
+{
+   // 1. Parse command line options
+   const char *mesh_file = "../data/star.mesh";
+   int order = 1;
+
+   OptionsParser args(argc, argv);
+   args.AddOption(&mesh_file, "-m", "--mesh", "Mesh file to use.");
+   args.AddOption(&order, "-o", "--order", "Finite element polynomial degree");
+   args.ParseCheck();
+
+   // 2. Read the mesh from the given mesh file, and refine once uniformly.
+   Mesh mesh(mesh_file);
+   mesh.UniformRefinement();
+
+   // 3. Define a finite element space on the mesh. Here we use H1 continuous
+   //    high-order Lagrange finite elements of the given order.
+   H1_FECollection fec(order, mesh.Dimension());
+   FiniteElementSpace fespace(&mesh, &fec);
+   cout << "Number of unknowns: " << fespace.GetTrueVSize() << endl;
+
+   // 4. Extract the list of all the boundary DOFs. These will be marked as
+   //    Dirichlet in order to enforce zero boundary conditions.
+   Array<int> boundary_dofs;
+   fespace.GetBoundaryTrueDofs(boundary_dofs);
+
+   // 5. Define the solution x as a finite element grid function in fespace. Set
+   //    the initial guess to zero, which also sets the boundary conditions.
+   GridFunction x(&fespace);
+   x = 0.0;
+
+   // 6. Set up the linear form b(.) corresponding to the right-hand side.
+   ConstantCoefficient one(1.0);
+   LinearForm b(&fespace);
+   b.AddDomainIntegrator(new DomainLFIntegrator(one));
+   b.Assemble();
+
+   // 7. Set up the bilinear form a(.,.) corresponding to the -Delta operator.
+   BilinearForm a(&fespace);
+   a.AddDomainIntegrator(new DiffusionIntegrator);
+   a.Assemble();
+
+   // 8. Form the linear system A X = B. This includes eliminating boundary
+   //    conditions, applying AMR constraints, and other transformations.
+   SparseMatrix A;
+   Vector B, X;
+   a.FormLinearSystem(boundary_dofs, x, b, A, X, B);
+
+   // 9. Solve the system using PCG with symmetric Gauss-Seidel preconditioner.
+   GSSmoother M(A);
+   PCG(A, M, B, X, 1, 200, 1e-12, 0.0);
+
+   // 10. Recover the solution x as a grid function and save to file. The output
+   //     can be viewed using GLVis as follows: "glvis -m mesh.mesh -g sol.gf"
+   a.RecoverFEMSolution(X, b, x);
+   x.SaveWithMesh("sol.gf", "mesh.mesh");
```

Comment:
I think this is slightly confusing, at first sight I thought this writes both the mesh and the solution in one file (reading such combined files was enabled a while back).

Reply:
Could this be resolved by changing the name to e.g. `SaveGridFunctionAndMesh`? (Which is more verbose, and `GridFunction:: SaveGridFunctionAndMesh` seems repetitive). Perhaps there is a different better name?

When I look at it, the fact that there are two file names is a pretty clear indication that it will be saved to two files.

Reply:
I guess from a software design point of view there should really be two statements, one saving the mesh and one saving the GridFunction, there is little reason to combine the two methods (unless we're saving one file). But that's just my opinion, if others are fine with this design it's not a huge deal (we should be pretty sure we want `SaveWithMesh` though, because it will stay in the API forever).

Reply:
I actually don't have a problem with the current name, but I do understand Jakub's concern.

I'm not sure we want to do this in Example 0, but another option could be to introduce a `DataCollection` and save that instead. At least in that case the meaning is clear, and maybe we should be encouraging the users to use the `DataCollection` classes more anyway?

Reply:
In an ideal world, the code would look like this IMO:
```
mesh.Save("mesh.mesh");
x.Save("sol.gf");
```
Doing it separately "disentangles" somewhat the two classes. `Print` would primarily be for printing to a stream (and you need to set the precision of the stream etc.), `Save` is for saving to a file, with sufficient precision. Right now `Mesh::Print` and `GF::Save` do the same thing, even though they have different names, unfortunately.

Reply:
I implemented this suggestion in #2130. Let me know what you think.

I actually like the solution in #2130 better than `SaveWithMesh`, so if the reviewers agree I will merge it here.

Reply:
One remaining question is what the default precision should be. I have it set to 8 right now because that is what is used in ex1. I almost think it should default to 16, so people won't be surprised if they lose a lot of digits of their solution, but that would double the file size…

Reply:
My two cents: Set it to 16. I always set it to 16 when I think of it, and disk space is cheap.